torchtune/modules/transformer.py

```diff
@@ -90,13 +99,147 @@ def forward(
         attn_out = self.attn(self.sa_norm(x), mask=mask, input_pos=input_pos)
 
         # Residual connection; shape: [batch_size, seq_length, embed_dim]
-        h = attn_out + x
+        h = self.sa_scale(attn_out) + x
+
+        # Norm applied before the feedforward layer
+        mlp_out = self.mlp(self.mlp_norm(h))
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        out = h + self.mlp_scale(mlp_out)
+        return out
+
+
+class TransformerCrossAttentionLayer(nn.Module):
+    """Cross attention Transformer layer following the same conventions as the TransformerSelfAttentionLayer.
+       Normalization is applied before the attention **and** FF layer.
+
+    Args:
+        attn (MultiHeadedAttention): Attention module.
+        mlp (nn.Module): Feed-forward module.
+        ca_norm (Optional[nn.Module]): Normalization to be applied before cross-attention.
+        mlp_norm (Optional[nn.Module]): Normalization to be applied before the feed-forward layer.
+        ca_scale (Optional[nn.Module]): Module to scale cross-attention output.
+        mlp_scale (Optional[nn.Module]): Module to scale the feed-forward output.
+    """
+
+    def __init__(
+        self,
+        attn: MultiHeadedAttention,
+        mlp: nn.Module,
+        *,
+        ca_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        ca_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
+    ) -> None:
+        super().__init__()
+        assert (
+            attn.pos_embeddings is None
+        ), "Doesn't support positional embeddings for cross attention, because q and k are different sequences."
```

Comment:
This would just be correct in the builder function as we'd be building a specific model. This is more a safety check for  if you're defining your own model and try to apply pos_embeddings naively to every layer. There may be a way to make it work with cross attention but I'd rather someone ran into this check and solved the problem instead of it silently happening incorrectly.